zerver/migrations/0443_backfill_subscription_auditlogs.py

```diff
@@ -0,0 +1,117 @@
+# Generated by Django 4.2 on 2023-04-19 18:18
+
+
+import logging
+from django.db import migrations
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Min, Max
+from django.utils.timezone import now as timezone_now
+
+
+def backfill_missing_subscriptions(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    """Backfill subscription realm audit log events for users which are
+    currently subscribed but don't have any, presumably due to some
+    historical bug.  This is important because those rows are
+    necessary when reactivating a user who is currently
+    soft-deactivated.
+
+    For each stream, we find the subscribed users who have no relevant
+    realm audit log entries, and create a backfill=True subscription
+    audit log entry which is the latest it could have been, based on
+    UserMessage rows.
+
+    """
+
+    Stream = apps.get_model("zerver", "Stream")
+    RealmAuditLog = apps.get_model("zerver", "RealmAuditLog")
+    Subscription = apps.get_model("zerver", "Subscription")
+    UserMessage = apps.get_model("zerver", "UserMessage")
+    Message = apps.get_model("zerver", "Message")
+
+    def get_last_message_id() -> int:
+        # We generally use this function to populate RealmAuditLog, and
+        # the max id here is actually system-wide, not per-realm.  I
+        # assume there's some advantage in not filtering by realm.
+        last_id = Message.objects.aggregate(Max("id"))["id__max"]
+        if last_id is None:
+            # During initial realm creation, there might be 0 messages in
+            # the database; in that case, the `aggregate` query returns
+            # None.  Since we want an int for "beginning of time", use -1.
+            last_id = -1
+        return last_id
+
+    for stream in Stream.objects.all():
+        subscribed_user_ids = set(
+            Subscription.objects.filter(recipient_id=stream.recipient_id).values_list(
+                "user_profile_id", flat=True
+            )
+        )
+        user_ids_in_audit_log = set(
+            RealmAuditLog.objects.filter(
+                realm=stream.realm,
+                event_type__in=[
+                    301,  # RealmAuditLog.SUBSCRIPTION_CREATED
+                    302,  # RealmAuditLog.SUBSCRIPTION_ACTIVATED
+                    303,  # RealmAuditLog.SUBSCRIPTION_DEACTIVATED
+                ],
+                modified_stream=stream,
+            )
+            .distinct("modified_user_id")
+            .values_list("modified_user_id", flat=True)
+        )
+
+        user_ids_missing_events = subscribed_user_ids - user_ids_in_audit_log
+        if not user_ids_missing_events:
+            continue
```

Comment:
Are we assuming that if a user has any of these events recorded, then that means they're surely not missing a `CREATED` event? Does that hold true? E.g. are there no users which with `DEACTIVATED` event, but missing the original `CREATED` event? 

Or is it just that it doesn't matter, and as long there is SOME latest event like that recorded, then it's fine and the soft reactivation path will be happy?

Reply:
[Done](https://chat.zulip.org/#narrow/stream/3-backend/topic/splitting.20out.20a.20development.20app/near/1557131).

Reply:
Ohhh, hang on.  Unsubscribing doesn't remove the Subscription, it just flips it to `active=False` -- I'd forgotten that column.

Reply:
Yeah. I think that means mean we need an "if not active` block to add a `DEACTIVATED` audit log entry, dated to the last message the user has in the stream?